ocw/lib/openstack.py

```diff
@@ -0,0 +1,108 @@
+        self.client()
+
+    def __new__(cls, vault_namespace: str):
+        if vault_namespace not in Openstack.__instances:
+            Openstack.__instances[vault_namespace] = self = object.__new__(cls)
+            self.__client = None
+        return Openstack.__instances[vault_namespace]
+
+    def client(self) -> None:
+        if self.__client is None:
+            self.__client = openstack.connect(
+                insecure=True,  # Trust the certificate
+                auth_url=self.get_data('auth_url'),
+                project_name=self.get_data('project_name'),
+                username=self.get_data('username'),
+                password=self.get_data('password'),
+                region_name=self.get_data('region_name'),
+                user_domain_name=self.get_data('user_domain_name'),
+                project_domain_name=self.get_data('project_domain_id'),
+            )
+        return self.__client
+
+    def cleanup_all(self) -> None:
+        self._cleanup_instances()
+        self._cleanup_images()
+        self._cleanup_keypairs()
+
+    def _cleanup_instances(self) -> None:
+        # Delete VM's & associated floating IP address(es)
+        try:
+            servers = [vm for vm in self.client().compute.servers() if vm.name.startswith("openqa-vm-")]
+        except OpenStackCloudException as exc:
+            self.log_warn("Got exception while listing instances: {}", exc)
+            return
+        self.log_dbg("Found {} servers", len(servers))
+        for server in servers:
+            if not self.is_outdated(parse(server.created_at)):
+                continue
+            if self.dry_run:
+                self.log_info("Instance termination {} skipped due to dry run mode", server.name)
+            else:
+                self.log_info("Deleting instance {}", server.name)
+                try:
+                    if not self.client().delete_server(
+                            server.name,
+                            wait=False,
+                            timeout=180,
+                            delete_ips=True,  # Delete floating IP address
+                            delete_ip_retry=1):
+                        self.log_err("Failed to delete instance {}", server.name)
+                except OpenStackCloudException as exc:
+                    self.log_warn("Got exception while deleting instance {}: {}", server.name, exc)
+
+    def _cleanup_images(self) -> None:
+        max_days = PCWConfig.get_feature_property('cleanup', 'openstack-max-age-days', self._namespace)
+        now = datetime.now(timezone.utc)
+        try:
+            images = [image for image in self.client().image.images() if "openqa" in image.tags]
+        except OpenStackCloudException as exc:
+            self.log_warn("Got exception while listing images: {}", exc)
+            return
+        self.log_dbg("Found {} images", len(images))
+        for image in images:
+            if (now - parse(image.created_at)).days > max_days:
```

Comment:
Why don't you use `is_outdated` as in servers?

Reply:
For VM's isn't 1 day enough?
For images it's configurable and 3 to 4 days is better to restart tests.

Reply:
> Why don't you use `is_outdated` as in servers?

You're right. It must be the same calculation done with VM's.

Reply:
we not discussing exact value here , we should have ability to tweak over a time when some requirements change . Only thing we discuss right now are we gonna do it via `max-age-days` or `openstack-max-age-days`  or  we keep some custom calculation within the code ( less preferable because than you can not actually tweak it )  ?

Reply:
> we not discussing exact value here , we should have ability to tweak it over a time when some requirements change . Only thing we discuss right now are we gonna do it via `max-age-days` or `openstack-max-age-days` or we keep some custom calculation within the code ( less preferable because than you can not actually tweak it ) ?

I prefer than for Openstack we have a different setting.

Also, do we actually see VM's running for more than one day?  For images the setting makes sense because tests can be restarted.

Reply:
> Also, do we actually see VM's running for more than one day?

I just want to have flexibility to give different answers to this question and not to have hardcoded value